### PR TITLE
[MRG] Update `MinHash.set_abundances` to remove hash if 0 abund; handle negative abundances.

### LINKED ISSUE #1355: support deletion of hashes by setting abundance to 0?
So https://github.com/dib-lab/sourmash/pull/1046 added a `clear` option to `set_abundances`, but setting abundances to zero doesn't delete the hash, which seems nonintuitive.

code:
```python
import sourmash
mh = sourmash.minhash.MinHash(n=0, ksize=31, scaled=1, track_abundance=True)

d = { 1: 5, 2: 3 }
mh.set_abundances(d)
print(dict(mh.hashes))

e = { 1: 0 }
mh.set_abundances(e, clear=False)
print(dict(mh.hashes))
```
yields
```
{1: 5, 2: 3}
{1: 5, 2: 3}
```

also what happens if there are negatives? you get an overflow error, which is ok but not intuitive :).

### PR DESCRIPTION
**Pull request by mr-eyes**
Update `MinHash.set_abundances(...)` to catch negative values and support hash removal by setting abundance to 0.

* This PR should add the following:
	* Remove the `hash` from the `sketch` if the abundance is set to zero.
	* Raise a `ValueError` exception if the abundance is a negative value.
* This PR Fixes #1355

### CODE REVIEW

File: src/sourmash/minhash.py [hunk 1]
```diff
@@ -625,7 +625,7 @@ def set_abundances(self, values, clear=True):
 
             for h, v in values.items():
                 hashes.append(h)
-                abunds.append(v)
+                abunds.append(max(0, v))
```
Comment (mr-eyes) [inline src/sourmash/minhash.py]: We can't cast negative values (signed int) to unsigned int to be passed to the [rust:add_hash_with_abundance](https://github.com/dib-lab/sourmash/blob/b42a9bedd153374df7851f2658698a4185eebc8a/src/core/src/sketch/minhash.rs#L313). Instead, convert the negative abundance values in Python to zero before passing to rust.

Comment (ctb) [inline src/sourmash/minhash.py]: Hot take: I think this behavior is strange! We don't support negative abundances, and probably shouldn't support passing in negative abundances.

How about:
* 0 deletes hash
* < 0 raises `ValueError`
?

File: tests/test_minhash.py [hunk 2]
```diff
@@ -1196,6 +1196,12 @@ def test_set_abundance_clear_4():
     a.set_abundances({20: 1, 10: 2}, clear=False)
     assert a.hashes == {10: 3, 20: 3}
 
+def test_clear_abundance_on_zero():
+    mh = sourmash.minhash.MinHash(n=0, ksize=31, scaled=1, track_abundance=True)
+    mh.set_abundances({ 1: 5, 2: 3, 3 : 5 })
+    mh.set_abundances({ 1: 0, 2 : -1 }, clear=False)
+    assert 1 not in dict(mh.hashes)
+    assert 2 not in dict(mh.hashes)
```
Comment (ctb) [inline tests/test_minhash.py]: should probably also check that the abundance of hash 3 is five, and length of mh.hashes is 1?

File: src/sourmash/minhash.py (hunk 1 above)
Comment (mr-eyes) [reply src/sourmash/minhash.py]: Yes, more reasonable.

Ok, I have a question. Is it useful/supported to change specific abundances without dropping the rest? `set_abundance` can add to the current values but can't change/reduce.

Comment (ctb) [reply src/sourmash/minhash.py]: Yah, I don't know 😄 . We aren't really using abundances a lot at the moment so I don't have many good use cases in mind. To me, that suggests that any choice we make is going to be wrong... It's easy enough to add new methods or adjust this one down the road, so let's just make sure whatever you choose is well tested! Unless @luizirber has thoughts.

Comment (mr-eyes) [reply src/sourmash/minhash.py]: hmmm... I was thinking of adding a `bool overwrite` flag so the user can overwrite some abundance values without clearing the rest, but, honestly, I have no idea about its importance. Let's just wait for a use case 😄

**Review by ctb — CHANGES_REQUESTED** [review]
Looks great! One last change request: could you update the docstring for set_abundances in Python, too?

**Comment by ctb** [comment]
Also, please correct the title and expand the description at the top of the PR by just a little bit to repeat the information that is currently in the title. Thanks!

**Comment by ctb** [comment]
nitpick: I'm going to change the title to mention `set_abundances` 😜

File: tests/test_minhash.py [hunk 3]
```diff
@@ -1196,6 +1196,20 @@ def test_set_abundance_clear_4():
...
+    mh.set_abundances({ 1: 5, 2: 3, 3 : 5 })
+    mh.set_abundances({ 1: 0 }, clear=False)
+    assert 1 not in dict(mh.hashes)
+    assert dict(mh.hashes)[2] == 3
+    assert dict(mh.hashes)[3] == 5
+    assert len(mh) == 2
+
+    with pytest.raises(ValueError):
+        mh.set_abundances({ 2: -1 }) # Test on clear = True
+        mh.set_abundances({ 2: -1 }, clear=False)
```
Comment (luizirber) [inline tests/test_minhash.py]: Should this be two separate `with` blocks? It is more code, but it is clearer that both statements should raise `ValueError` (if only one raises the exception the test still passes)

File: src/sourmash/minhash.py (hunk 1 above)
Comment (luizirber) [reply src/sourmash/minhash.py]: :+1: on waiting for use case

File: tests/test_minhash.py (hunk 3 above)
Comment (mr-eyes) [reply tests/test_minhash.py]: I got it. Fixed now. Thank you!

**Review by ctb — APPROVED** [review]

**Pull request merged**